mitxgraders/helpers/math_array.py

```diff
@@ -0,0 +1,371 @@
+from __future__ import division # necessary for one of the doctests
+import numpy as np
+from numbers import Number
+from mitxgraders.baseclasses import StudentFacingError
+from mitxgraders.helpers.mathfunc import robust_pow
+
+class MathArrayError(StudentFacingError):
+    """
+    Thrown by MathArray when anticipated errors are made.
+    """
+
+def is_number_zero(value):
+    """
+    Tests whether a value is the scalar number 0.
+
+    >>> map(is_number_zero, [0, 0.0, 0j])
+    [True, True, True]
+    >>> is_number_zero(np.matrix([0, 0, 0]))
+    False
+    """
+    return isinstance(value, Number) and value == 0
+
+def is_scalar_array_zero(value):
+    return isinstance(value, MathArray) and value.ndim == 0 and value.item() == 0
+
+def is_square(array):
+    return array.ndim == 2 and array.shape[0] == array.shape[1]
+
+class MathArray(np.ndarray):
+    """
+    A modification of numpy's ndarray class that behaves more like a mathematician would expect.
+
+    Terminology:
+    ============
+        - vector, 1 dimensional, e.g., MathArray([1, 2, 3])
+        - matrix, 2 dimensional, e.g., MathArray([[1, 2], [3, 4]])
+        - tensor, 3+ dimensional, e.g., MathArray([ [[1]], [[2]] ])
+
+    Zero-dimensional arrays, e.g., MathArray(5) are also supported but should
+    not occur in practice.
+
+        - scalar: Python number OR zero-dimensional array
+
+    Differences from numpy.ndarray:
+    ==============================
+        - Multiplication:
+            - scalar * array = scales elements of array;
+            - array * scalar = scales elements of array;
+            If shapes are correct:
+            - vector * vector = number
+            - matrix * vector = vector
+            - vector * vector = vector
+            - matrix * matrix = matrix
+            Any multiplication involving tensors and arrays of dimension >1 is not supported.
+        - powers are only allowed for square matrices and integer-like exponents.
```

Comment:
non-negative integer-like?

Reply:
Negative integers are fine if the matrix is numerically invertible.

Actually, that's worth thinking about more. Maybe there should be a matrix sampling class that ensures invertibility of the sample matrices.

Reply:
Waaait, numpy will do a matrix inversion automatically if you give it a negative power? Duuude!

I'm tempted to turn this off and maybe expose a function inv() to invert a matrix. Otherwise, matrix inversion problems will be trivially defeated, with no way to disallow the inverse operation. If inv is a function, the function can be blacklisted.

Reply:
Also, there are so many matrix types to sample, we're going to have a field day writing samplers for them all ;-)

Reply:
> I'm tempted to turn this off and maybe expose a function inv() to invert a matrix. Otherwise, matrix inversion problems will be trivially defeated, with no way to disallow the inverse operation. If inv is a function, the function can be blacklisted.

That's a good point. I have not made that change, but I did add two tests to assert that zero and negative powers work as (currently) expected, just so that I remember to document and test the behavior if we do change it.